sound/soc/sof/pm.c

```diff
@@ -431,6 +431,36 @@ int snd_sof_set_d0_substate(struct snd_sof_dev *sdev,
 }
 EXPORT_SYMBOL(snd_sof_set_d0_substate);
 
+/*
+ * Audio DSP states may transform as below:-
+ *
+ *                                         D0I3-compatible stream
+ *     Runtime    +----------------------+  opened only, timeout
+ *     suspend    |                      +------------------------+
+ *   +------------+       D0(active)     |                        |
+ *   |            |                      <------------------+     |
+ *   |   +-------->                      |   The last D0I3  |     |
+ *   |   |Runtime +--^--+----------^--+--+   stream closed  |     |
+ *   |   |resume     |  |          |  |                +----------v-----+
+ *   |   |           |  |          |  |                | run-time D0I3  |
+ * +-v---+---+ resume|  |          |  |                |    (active)    |
+ * |   D3    |  from |  | D3       |  |d0i3_suspend    +-------+--------+
+ * |run-time |   D3  |  |suspend   |  |                        |
+ * |suspended|       |  |          |  |                        |d0i3_suspend
+ * +---------+       |  |          |  |                        |
+ *                   |  |          |  |                        |
+ *             +-----+--v-------+  |  |             +----------v-----+
+ *             |                |  |  +------------->                |
+ *             |  D3 suspended  |  |                | D0I3 suspended |
+ *             |                |  +----------------+                |
+ *             |                |    resume from    |                |
+ *             +----------------+    d0i3_suspend   +----------------+
+ *
+ *
+ * d0i3_suspend = s0_suspend && D0I3 stream opened,
+ * D3 suspend = !d0i3_suspend,
+ */
+
```

Comment:
@keyonjie in my interpretation there are only 3 states for the DSP (D0, D0i3 and D3) So it is a bit confusing when your say D3 runtime-suspended and D3 suspended. Its not quite clear what it really means. 

I made a simpler diagram. Let me know if you cant understand my handwriting. I may have missed something but I think this shows the transitions between the DSP states better. 
![IMG_20191104_161010](https://user-images.githubusercontent.com/7766921/68168558-076a0080-ff1e-11e9-83a7-0e04ae4a12f0.jpg)

Reply:
The above diagram only applies if the platform supports D0i3. The other thing is this diagram is intel specific. So needs to be moved there

Reply:
> @keyonjie in my interpretation there are only 3 states for the DSP (D0, D0i3 and D3) So it is a bit confusing when your say D3 runtime-suspended and D3 suspended. Its not quite clear what it really means.
> 
> I made a simpler diagram. Let me know if you cant understand my handwriting. I may have missed something but I think this shows the transitions between the DSP states better.
> ![IMG_20191104_161010](https://user-images.githubusercontent.com/7766921/68168558-076a0080-ff1e-11e9-83a7-0e04ae4a12f0.jpg)

Thanks, basically, I think your diagram is good to describe to DSP state transitions, except that it doesn't include:
1. the runtime PM ops handling (the left-upper part in my diagram), we can treat the DSP state in runtime-suspended as a sub-state of D0 in your diagram, and don't care about its internal transition?
2. the difference of S0 D0i3 and S0Ix D0i3. In your diagram, you try to include both in D0<-->D0i3, but it doesn't describe how should we handle when the S0Ix suspend arrive if we are already in S0 D0i3, which is we need to do here in pm.c:snd_sof_suspend/resume(). Yes, again, we can treat both S0-D0i3 and S0Ix-D0i3 as sub-states of D0i3 and ignore its internal transition here.

Then should we draw another diagram to describe the internal transition for '2' and '1'?

Reply:
I knew I missed something and yes the runtime PM case is important. Let me add it and send you a new diagram  in a bit

Reply:
agree with Ranjani, there is no difference between runtime_suspended D3 and system_suspended D3. it's D3.

Reply:
> agree with Ranjani, there is no difference between runtime_suspended D3 and system_suspended D3. it's D3.

@plbossart then how do you suggest to add the runtime_suspend/resume handling here?
If we are in D3 via runtime_suspend, we can't resume it from a S3->S0 or S0Ix->S0. That's why they are actually different state to me.

Reply:
![IMG_20191104_180055](https://user-images.githubusercontent.com/7766921/68172718-70a54000-ff2d-11e9-89ba-4e42378dc10a.jpg)
@keyonjie how about this one? It handles the runtime PM case as well.
The important thing to remember is that before system suspend, the audio device will always be runtime-resumed, so it will be in D0.

Reply:
> the audio device will always be runtime-resumed, so it will be in D0.

the audio device will be runtime-resumed if needed. This is the default but you can decide to go directly to system suspend. For SoundWire we don't pm_runtime resume, it's possible but not a requirement.

Reply:
I also wonder if we absolutely need to have the arrow at the bottom. Can we actually go from S0ix to S3 without going to S0 first?

Reply:
oh the arrow at the bottom is only for S0 -> S3. The DSP can be in D0i3 in S0. So when transitioning to S3, the DSP will need to be put in D3.

Reply:
> the audio device will be runtime-resumed if needed. This is the default but you can decide to go directly to system suspend. For SoundWire we don't pm_runtime resume, it's possible but not a requirement.
@plbossart Im not too familiar with what happens in SDW but the audio DSP will always be runtime resumed before entering S0ix/S3 isnt it?

Reply:
> oh the arrow at the bottom is only for S0 -> S3. The DSP can be in D0i3 in S0. So when transitioning to S3, the DSP will need to be put in D3.

Hi @ranj063 Thanks for update, it almost looks good, but here when the DSP in D0i3 in S0, it need to be put in D3 for S0->S3 transition, or to be put in D0i3(+ IRQ wakeup + disable dma trace) for S0->S0ix transition.

Reply:
> Hi @ranj063 Thanks for update, it almost looks good, but here when the DSP in D0i3 in S0, it need to be put in D3 for S0->S3 transition, or to be put in D0i3(+ IRQ wakeup + disable dma trace) for S0->S0ix transition.

@keyonjie right, but there is no DSP state transition for the S0->S0ix case. Disabling trace dma and enabling IRQ wake-up should be handled by the platform-specific suspend routine.

Reply:
um, it depends on how do you define the DSP state, e.g. we may put DSP to lower power state D0i3 when the system is in S0ix suspended, comparing with the S0 D0i3, so the flags in PM_GATE IPC will be different for them.

I am fine to use the simpler diagram here to make it easier for understanding, but we need to keep in mind that we still have difference scenarios(or called sub-state) inside both the D3 and D0i3 state, when we are in coding of the detail logic handling there.

Reply:
> um, it depends on how do you define the DSP state, e.g. we may put DSP to lower power state D0i3 when the system is in S0ix suspended, comparing with the S0 D0i3, so the flags in PM_GATE IPC will be different for them.

@keyonjie Why would it be different in S0? Why cant you put the DSP in the same lower power D0i3 in S0?

Reply:
> > um, it depends on how do you define the DSP state, e.g. we may put DSP to lower power state D0i3 when the system is in S0ix suspended, comparing with the S0 D0i3, so the flags in PM_GATE IPC will be different for them.
> 
> @keyonjie Why would it be different in S0? Why cant you put the DSP in the same lower power D0i3 in S0?

For example, per my observation, in S0 D0i3, access to memory windows is possible, and we are planning to keep DMA trace active in this state, for debug purpose, but when the system is entering S0ix, everything is changed, I don't have enough knowledge to explain how PMC will handling to the audio domain on S0ix entering, but I am quite sure the behavior of the Audio DSP in this S0ix D0i3 has difference with the one in S0 D0i3.

And, from the system point of view, we might need more normal interaction between FW & host in S0 D0i3 without putting the DSP back to D0, which is impossible in S0ix D0i3. When we are in S0ix D0i3, the only interaction between DSP and host is that the FW will send an IPC to wake the host up.

Reply:
Then you have different definitions of D0i3....
You should have one well defined behavior. If you have multiple behaviors, then you are not validating the same things.
I think the memory windows and DMA should only be allowed in D0i3_streaming. @mmaka1 and @lbetlej mentioned this state when we started this, and maybe it's time to formally recognize the difference between usages.

Reply:
> Then you have different definitions of D0i3....
> You should have one well defined behavior. If you have multiple behaviors, then you are not validating the same things.
> I think the memory windows and DMA should only be allowed in D0i3_streaming. @mmaka1 and @lbetlej mentioned this state when we started this, and maybe it's time to formally recognize the difference between usages.

Oh yes, that is a major scenario of S0 D0i3, I think.